ci/ci/build.py

```diff
@@ -89,15 +89,18 @@ def __init__(self, code, config_str, scope, requested_step_names=None):
         for step_config in config['steps']:
             step_params = StepParameters(code, scope, step_config, name_step)
             step = Step.from_json(step_params)
-            self.steps.append(step)
-            name_step[step.name] = step
+            if not step.run_if_requested or step.name in requested_step_names:
+                self.steps.append(step)
+                name_step[step.name] = step
+            else:
+                name_step[step.name] = None
 
-        # transitively close requested_step_names over dependenies
+        # transitively close requested_step_names over dependencies
         if requested_step_names:
             visited = set()
 
             def request(step):
-                if step not in visited:
+                if step and step not in visited:
```

Comment:
OK, almost done: this change is no longer needed, because step should be non-None.  deps are filtered to be non-None, and requested steps cannot be None.